cortipy/cortical_client.py

```diff
@@ -134,13 +83,66 @@ def __init__(self,
     else:
       self.apiUrl = baseUrl
 
+    if useCache and not os.path.exists(cacheDir):
+      os.makedirs(cacheDir)
+
+    self.useCache = useCache
     self.cacheDir = cacheDir
     self.verbosity = verbosity
     self.retina = retina
-    self.useCache = useCache
+    self._session = requests.Session()
+
+
+  def _cachedRequest(self, fn, url, params, headers, data=None):
+    def _doRequest():
+      # Internal request wrapper to
+      extras = {}
+      if data:
+        extras["data"] = data
+
+      response = fn(url, params=params, headers=headers, **extras)
+
+      if response.status_code != 200:
+        raise UnsuccessfulEncodingError(
+          "Response " + str(response.status_code) + ": " + response.content)
+
+      try:
+        responseObj = json.loads(response.content)
+      except ValueError("Could not decode the query response."):
+        responseObj = []
+
+      return responseObj
+
+    if not self.useCache:
+      return _doRequest()
+
+    # Construct deterministic hash for request
+
+    m = hashlib.sha224()
+
+    m.update(fn.__name__)
+    m.update(url)
+    m.update("".join(str(y) for x in sorted(headers.iteritems()) for y in x))
+    m.update("".join(str(y) for x in sorted(params.iteritems()) for y in x))
+
+    if data:
+      m.update(data)
+
+    cacheKey = m.hexdigest()
+
+    cachePath = os.path.join(self.cacheDir, "{}.msgpack".format(cacheKey))
+
+    if os.path.isfile(cachePath):
+      # Request has been made before, load and return original response
+      response = msgpack.load(open(cachePath, "r"))
+    else:
+      # Make new request, cache the response
+      response = _doRequest()
```

Comment:
What happens if CIO API was unable to fulfill this request or if they return an empty bitmap? Shouldn't we generate a placeholder bitmap here?